pkg/netns/netns_linux.go

```diff
@@ -0,0 +1,200 @@
+// UNSHARE syscall - the original network namespace will be retained and a handle to the
+// new network namespace will be returned.
+//
+// The namespace created by this call is NOT pinned, or bind-mounted, and so will be
+// closed/lost when the process terminates.
+//
+// If you need a pinned network namespace, call `CreateNewNetns` with pin == true, or call Pin() on
+// the returned handle.
+func New() (NetNSHandle, error) {
+	var innerError error
+	var newNetns NetNSHandle
+	restore, err := saveRestoreCurrentThreadNetns()
+	if err != nil {
+		return newNetns, err
+	}
+	defer restore()
+
+	// this invokes the UNSHARE syscall, which creates a new netns *and moves the current thread*
+	// to that netns. So, after creating and getting a FD, we need to move it BACK (via defer), so that
+	// it looks like we've never left the original netns.
+	fd, err := createAndMoveToNewNetNS()
+	if err != nil {
+		return NetNSHandle{fd: -1}, fmt.Errorf("Unable to create new netns: %w", err)
+	}
+	return NetNSHandle{fd: int(fd)}, innerError
+}
+
+// DoByPinPath is for when a process (CNI plugins, etc) needs to jump into a netns
+// managed/owned by a different process in order to do things, and that other process
+// has "pinned" or bind-mounted the netns to a file.
+//
+// Cilium code should not pin or bind-mount network namespaces it creates, this function is
+// purely for interacting with network namespaces created and managed outside of Cilium.
+func DoByPinPath(netNSPath string, runInside func(*NetNSHandle) error) error {
+	fd, err := unix.Open(netNSPath, unix.O_RDONLY|unix.O_CLOEXEC, 0)
+	if err != nil {
+		return err
+	}
+	// We aren't keeping this and we don't own the pin or the netns it refers to
+	// so defer closure.
+	defer func() {
+		err = errors.Join(err, unix.Close(fd))
+	}()
+
+	localHandle := NetNSHandle{
+		fd: int(fd),
+	}
+	err = localHandle.Do(runInside)
+	return err
+}
+
+// FD returns the underlying file descriptor of the NetNSHandle
+// Note that FDs should not be used for comparing or testing equality, for that
+// see `equals`
+func (h *NetNSHandle) FD() int {
+	return h.fd
+}
+
+// Close closes the handle to the network namespace. Note that this does not guarantee
+// the network namespace itself will be destroyed or closed, which will happen when all references to it
+// (including the one being closed here) are gone.
+func (h *NetNSHandle) Close() error {
+	if err := unix.Close(h.fd); err != nil {
+		return err
+	}
+	h.fd = -1
+	return nil
+}
+
+// Do runs the provided func within a new thread+netns, without changing the current thread's netns
+//
+// Note that since network namespaces are set *per thread*, the thread the closure is running inside
+// will be locked with `runtime.LockOSThread()` during this operation.
+//
+// This means the provided closure should _not_ call runtime.UnlockOSThread() at all,
+// or else undefined behavior should be expected.
+func (h *NetNSHandle) Do(runInside func(*NetNSHandle) error) error {
```

Comment:
```suggestion
func (h *NetNSHandle) Do(f func() error) error {
```

As far as I can see, the `*NetNSHandle` parameter isn't used anywhere. I think I've mentioned this in a previous comment, but I don't think we should ever pass the handle. Close() seems a little pointless to call from within the Do() closure, and recursive calls to Do() are absolutely not safe.

A secondary benefit to removing the NetNSHandle argument from function passed to Do() is that we should then be able to unexport `NetNSHandle` by renaming it to `netNS`. Callers should never synthesize their own netns handles, as they wouldn't contain a finalizer.

Also nit: functions passed as arguments are commonly just called `f`.

Reply:
> As far as I can see, the *NetNSHandle parameter isn't used anywhere. I think I've mentioned this in a previous comment, but I don't think we should ever pass the handle. 

Yes, you did, and thanks for reminding me again - I twiddled the existing code in `plugins/cilium-cni/cmd/cmd.go` that led me to include it in the first place, so we don't need it at all anymore, will remove.

> recursive calls to Do() are absolutely not safe.

AFAIK it should be fine in the sense that we would as a result have multiple threads in the same netns and aside from eventually exhausting the thread pool, that's fine in a Linux/go context, but I might be missing something.

Either way, will drop it since it's not needed.

> Also nit: functions passed as arguments are commonly just called f.

Yes, it's a common convention I'm not fond of, but will fix it :D

Reply:
Fixed this, missing that it was totally unused now was a bonehead move on my part, apologies for making you repeat yourself.